test/ieee754_linear_test.cpp

```diff
@@ -0,0 +1,80 @@
+//  (C) Copyright Ryan Elandt 2023.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#define BOOST_TEST_MAIN
+#include <boost/test/unit_test.hpp>
+#include <boost/test/tools/floating_point_comparison.hpp>
+#include <boost/test/results_collector.hpp>
+
+#include <boost/multiprecision/cpp_bin_float.hpp>
+#include <boost/multiprecision/cpp_dec_float.hpp>
+#include <boost/multiprecision/cpp_complex.hpp>
+#include <boost/math/concepts/real_concept.hpp>
+
+#ifdef BOOST_HAS_FLOAT128
+#include <boost/multiprecision/float128.hpp>
+#endif
+
+#if __has_include(<stdfloat>)
```

Comment:
It's unfortunate, but __has_include is an insufficient check, because it may pass even when the translation mode is not C++23, so we will need to check the value of BOOST_CXX_VERSION as well (unfortunately __cplusplus is non-reliable for MSVC).  I yes, must get around to adding proper Boost.Config macros and tests for these new C++23 headers sometime!

Reply:
The include pattern:
```c++
#if __has_include(<stdfloat>)
#  include <stdfloat>
#endif
```
occurs in `include/boost/math/tools/promotion.hpp`, `include/boost/math/concepts/real_concept.hpp`, and about 30 test files including `test/jacobi_test.cpp`. I think it makes sense to allow it here for consistency. If required it should be fixed across the codebase in a separate PR.

Reply:
This came up here: https://github.com/boostorg/math/issues/1023 and the headers were fixed at that time.  I hadn't realised this had been sprinkled across all the tests as well :(

The only remaining header I can see which is incorrect is here: https://github.com/boostorg/math/blob/09d82da6e9884da86495279b331ff038dd1330b8/include/boost/math/ccmath/signbit.hpp#L28  @mborland ?

Reply:
It might be desirable for the pattern:
```c++
#if defined __has_include
#  if __cplusplus > 202002L || _MSVC_LANG > 202002L 
#    if __has_include (<X>)
#    include <X>
#    endif
#  endif
#endif
```
to me made into a macro similar to
```c++
INCLUDE_CPP23_HEADER(X)
```

Reply:
Yes, but you still need to be able to determine whether the header has been included or not (ie whether you can use those contents or not), so the normal method is:

```
#ifndef BOOST_NO_CXX23_STDFLOAT
#include <stdfloat>
#endif 



#ifndef BOOST_NO_CXX23_STDFLOAT

// do something that uses the header

#endif 
```

I'm working on the C++23 macros for Boost.Config now...

Reply:
The approach described above typically applies when one is manually defining methods. A manually defined method may compile if certain headers have been included. For example, one might manually define a method for type `__float128`, another for `_Quad`, and another for `std::float128_t`. Careful preprocessor use is required because certain methods will only compile if the correct headers have been included.

In contrast to this approach, this approach relies on duck typing. In C++ duck typing happens at compile time and is implemented with templates and SFINAE. It also allows a single templatized method to be used for various types. For example, the 128 bit type is used by the following types: `__float128`, `_Quad`, `std::float128_t`, `multiprecision::float128`, as well as `std_real_concept` when it's set up to emulate a 128 bit float. I may be missing some, but the compiler will not.

The benefit of the duck type approach that _the code only needs to compile for the template parameters that are actually used_. For this reason, the approach does not require the careful preprocessor use of the manual approach. 

Also, thank you for working on the C++23 macros.